src/services/identity/UsersService.ts

```diff
@@ -155,26 +155,46 @@ class UsersService {
 
   async canSendEmailOtp(email: string) {
     const normalizedEmail = email.toLowerCase()
-    const whitelistEntries = await this.whitelist.findAll({
-      attributes: ["email"],
-      where: {
-        expiry: {
-          [Op.or]: [{ [Op.is]: null }, { [Op.gt]: new Date() }],
+
+    // raw query because ORMs suck!
+    // why: we want to leverage the DB to see if a single record exists that whitelists the input
+    // we do not want to download the whole table locally to filter in local code
+    // query logic:
+    // - if whitelist entry is a full email (something before the @), then do exact match
+    // - if whitelist entry is a domain (no @, or starting with @), then do suffix match
+    // Limit 1 is added to allow the query to exit early on first match
+    const records = (await this.sequelize.query(
+      `
+        SELECT email
+        FROM whitelist
+        WHERE
+          (expiry is NULL OR expiry >= NOW())
+          AND
+          CASE WHEN email ~ '^.+@'
```

Comment:
Note: compared to the previous code, this is a simpler regexp. It detects the "full email", while the previous more complicated regexp was detecting the domains. Obviously the match strategies have been swapped accordingly.